geode/src/main/java/org/apache/zeppelin/geode/GeodeOqlInterpreter.java

```diff
@@ -0,0 +1,297 @@
+ * <li>{@code geode.locator.port} - The Geode Locator {@code <PORT>} to connect to.</li>
+ * </ul>
+ * <p>
+ * Sample usages: <br/>
+ * {@code %geode.oql} <br/>
+ * {@code SELECT * FROM /regionEmployee e WHERE e.companyId > 95} <br/>
+ * {@code SELECT * FROM /regionEmployee ORDER BY employeeId} <br/>
+ * {@code 
+ * SELECT * FROM /regionEmployee 
+ * WHERE companyId IN SET(1, 3, 7) OR lastName IN SET('NameA', 'NameB')
+ * } <br/>
+ * {@code 
+ * SELECT e.employeeId, c.id as companyId FROM /regionEmployee e, /regionCompany c 
+ * WHERE e.companyId = c.id
+ * }
+ * </p>
+ * <p>
+ * OQL specification and sample queries:
+ * http://geode-docs.cfapps.io/docs/getting_started/querying_quick_reference.html
+ * </p>
+ * <p>
+ * When the Zeppelin server is collocated with Geode Shell (gfsh) one can use the %sh interpreter to
+ * run Geode shell commands: <br/>
+ * {@code 
+ * %sh
+ *  source /etc/geode/conf/geode-env.sh
+ *  gfsh << EOF
+ *    connect --locator=ambari.localdomain[10334]
+ *    destroy region --name=/regionEmployee
+ *    create region --name=regionEmployee --type=REPLICATE
+ *    exit;
+ *  EOF
+ *}
+ * </p>
+ * <p>
+ * Known issue:http://gemfire.docs.pivotal.io/bugnotes/KnownIssuesGemFire810.html #43673 Using query
+ * "select * from /exampleRegion.entrySet" fails in a client-server topology and/or in a
+ * PartitionedRegion.
+ * </p>
+ */
+public class GeodeOqlInterpreter extends Interpreter {
+
+  private Logger logger = LoggerFactory.getLogger(GeodeOqlInterpreter.class);
+
+  private static final String DEFAULT_PORT = "10334";
+  private static final String DEFAULT_HOST = "localhost";
+
+  private static final char NEWLINE = '\n';
+  private static final char TAB = '\t';
+  private static final char WHITESPACE = ' ';
+
+  private static final String TABLE_MAGIC_TAG = "%table ";
+
+  public static final String LOCATOR_HOST = "geode.locator.host";
+  public static final String LOCATOR_PORT = "geode.locator.port";
+
+  static {
+    Interpreter.register("oql", "geode", GeodeOqlInterpreter.class.getName(),
+        new InterpreterPropertyBuilder().add(LOCATOR_HOST, DEFAULT_HOST, "The Geode Locator Host.")
+            .add(LOCATOR_PORT, DEFAULT_PORT, "The Geode Locator Port").build());
+  }
+
+  private ClientCache clientCache = null;
+  private QueryService queryService = null;
+  private Exception exceptionOnConnect;
+
+  public GeodeOqlInterpreter(Properties property) {
+    super(property);
+  }
+
+  protected ClientCache getClientCache() {
+
+    String locatorHost = getProperty(LOCATOR_HOST);
+    int locatorPort = Integer.valueOf(getProperty(LOCATOR_PORT));
+
+    ClientCache clientCache =
+        new ClientCacheFactory().addPoolLocator(locatorHost, locatorPort).create();
+
+    return clientCache;
+  }
+
+  @Override
+  public void open() {
+    logger.info("Geode open connection called!");
+    try {
+      clientCache = getClientCache();
+      queryService = clientCache.getQueryService();
```

Comment:
Is it possible/allowed to call `#open()` more than once?
Because if it is possible then it will leak resources (ClientCache and QueryService).
Maybe it is a good idea to always start with a call to `close()` before creating new resources?